File: frontend/src/lib/tool-view.ts
import type { JsonSchema, ToolSpec, UpstreamTool } from './types'

/**
 * What the agent sees for one curated tool: alias, rewritten description, and
 * the schema with presets and hidden arguments removed. Mirrors the backend.
 */
export function agentFacingSchema(spec: ToolSpec, upstream: UpstreamTool | undefined): JsonSchema {
  const base: JsonSchema = JSON.parse(JSON.stringify(upstream?.inputSchema ?? { type: 'object', properties: {} })) as JsonSchema
  base.type = 'object'
  base.properties = base.properties ?? {}
  const removed = new Set([...Object.keys(spec.presets), ...spec.hidden_args])
  for (const k of removed) delete base.properties[k]
  if (base.required) {
    base.required = base.required.filter((r) => !removed.has(r))
    if (!base.required.length) delete base.required
  }
  return base
}

export function agentFacingTool(spec: ToolSpec, upstream: UpstreamTool | undefined) {
  return {
    name: spec.alias,
    description: spec.description_override ?? upstream?.description ?? '',
    inputSchema: agentFacingSchema(spec, upstream),
  }
}

export function estimateTokens(text: string) {
  return Math.round(text.length / 4)
}

File: backend/app/services/tool_view.py
"""What the agent sees. Pure functions, mirrored by frontend/src/lib/tool-view.ts."""

from __future__ import annotations

import copy
from typing import Any

Manifest = dict[str, dict[str, Any]]  # server_id -> {upstream_name -> tool}


def index_manifests(servers: list[dict[str, Any]]) -> Manifest:
    out: Manifest = {}
    for s in servers:
        tools = (s.get("manifest") or {}).get("tools", [])
        out[s["_id"]] = {t["name"]: t for t in tools}
    return out


def exposed_schema(spec: dict[str, Any], upstream: dict[str, Any] | None) -> dict[str, Any]:
    """Upstream inputSchema minus presets and hidden arguments, with `required` fixed."""
    schema = copy.deepcopy((upstream or {}).get("inputSchema") or {"type": "object", "properties": {}})
    schema["type"] = "object"
    props = schema.get("properties")
    if not isinstance(props, dict):
        props = {}
    schema["properties"] = props
    removed = set(spec.get("presets") or {}) | set(spec.get("hidden_args") or [])
    for key in removed:
        props.pop(key, None)
    required = schema.get("required")
    if isinstance(required, list):
        kept = [r for r in required if r not in removed]
        if kept:
            schema["required"] = kept
        else:
            schema.pop("required", None)
    return schema


def exposed_tool(spec: dict[str, Any], upstream: dict[str, Any] | None) -> dict[str, Any]:
    description = spec.get("description_override")
    if description is None:
        description = (upstream or {}).get("description") or ""
    tool: dict[str, Any] = {
        "name": spec["alias"],
        "description": description,
        "inputSchema": exposed_schema(spec, upstream),
    }
    if upstream and upstream.get("outputSchema"):
        tool["outputSchema"] = upstream["outputSchema"]
    return tool


def build_tools(loadout: dict[str, Any], manifests: Manifest) -> list[dict[str, Any]]:
    """tools/list for a loadout: enabled specs whose upstream still exists, in spec order."""
    out: list[dict[str, Any]] = []
    for spec in loadout.get("tools", []):
        if not spec.get("enabled", True):
            continue
        upstream = manifests.get(spec["server_id"], {}).get(spec["upstream_name"])
        if upstream is None:
            continue
        out.append(exposed_tool(spec, upstream))
    return out


def outbound_arguments(spec: dict[str, Any], incoming: dict[str, Any]) -> tuple[dict[str, Any], list[str]]:
    """Strip hidden arguments the agent sent anyway, then let presets win."""
    hidden = set(spec.get("hidden_args") or [])
    stripped = [k for k in incoming if k in hidden]
    args = {k: v for k, v in incoming.items() if k not in hidden}
    args.update(spec.get("presets") or {})
    return args, stripped
